File: bob-server/src/fr/dmconcept/bob/server/ioio/ProjectLock.scala
package fr.dmconcept.bob.server.ioio;

import android.util.Log
import fr.dmconcept.bob.server.activities.MainIOIOActivity

import java.util.ArrayList
import fr.dmconcept.bob.client.models.Project

class ProjectLock {

  var project: Project = null
  var duration : Int = 0
  var sliceCount : Int = 0
  var slices = new java.util.ArrayList[java.util.ArrayList[Int]]()

  def setProject(p: Project) {
    project    = p
    duration   = projectDuration
    sliceCount = this.duration / (MainIOIOActivity.PERIOD / 1000)
    slices     = sliceSteps()
  }

  private def projectDuration: Int =  {

    var duration = 0

    for (step <-  project.steps)
      duration = duration + step.duration

    duration

  }

  private def sliceSteps() : java.util.ArrayList[java.util.ArrayList[Int]] = {

    val s = new java.util.ArrayList[java.util.ArrayList[Int]]

    for (i <-  0 to (project.steps.length - 1)) {

      val step     = project.steps(i    )
      val nextStep = project.steps(i + 1)

      val stepSliceCount: Int = step.duration / (MainIOIOActivity.PERIOD / 1000)

      Log.i("~~~~~~~~~~~~~~~~", s"Step duration=${step.duration} ms, slices=$stepSliceCount")

      for (sliceIndex <- 0 to stepSliceCount) {

        val interpolatedPositions = new ArrayList[Int]()

        for (positionIndex <-  0 to step.positions.length) {

          val initialPosition = step.positions(positionIndex)
          val endPosition     = nextStep.positions(positionIndex)

          // Interpolate the position
          val deltaPosition : Float = endPosition - initialPosition
          val deltaPerSlice : Float = deltaPosition / stepSliceCount

          //                    if (sliceIndex == 0)
          //                        Log.i("~~~~~~~~~~~~~~~~", "S" + positionIndex + " [" + initialPosition + " -> " + endPosition + "] delta: " + deltaPosition + ", deltaPerSlice: " + deltaPerSlice);

          val interpolatedPercentage : Float = initialPosition + deltaPerSlice * sliceIndex
          val interpolatedPosition   : Float = MainIOIOActivity.MIN + interpolatedPercentage * MainIOIOActivity.PERCENT

          interpolatedPositions.add(interpolatedPosition.intValue())

        }

        if (sliceIndex % 10 == 0)
          Log.i("~~~~~~~~~~~~~~~~", s"slice $sliceIndex: $interpolatedPositions")

        s.add(interpolatedPositions)
      }

      Log.i("~~~~~~~~~~~~~~~~", "")

    }

    if (s.size() != sliceCount)
      throw new RuntimeException("Expecting " + sliceCount + " slices but got " + s.size())

    return s

  }

}

File: bob-server_gradle/app/src/main/java/fr/dmconcept/bob/server/ioio/ProjectLock.java
package fr.dmconcept.bob.server.ioio;

import android.os.SystemClock;
import android.util.Log;
import fr.dmconcept.bob.server.activities.MainIOIOActivity;
import fr.dmconcept.bob.server.models.Project;
import fr.dmconcept.bob.server.models.Step;

import java.util.ArrayList;

public class ProjectLock {

    protected Project project;
    protected int duration;
    protected int sliceCount;
    protected ArrayList<ArrayList<Integer>> slices = new ArrayList();

    public ProjectLock() {

    }

    public Project getProject() {
        return project;
    }

    public int getDuration() {
        return duration;
    }

    public int getSliceCount() {
        return sliceCount;
    }

    public ArrayList<ArrayList<Integer>> getSlices() {
        return slices;
    }

    public void setProject(Project p) {
        this.project    = p;
        this.duration   = projectDuration();
        this.sliceCount = this.duration / (MainIOIOActivity.PERIOD / 1000);
        this.slices     = sliceSteps();
    }

    private int projectDuration() {

        int duration = 0;
        for (Step step: project.steps)
            duration += step.duration;

        return duration;
    }

    private ArrayList<ArrayList<Integer>> sliceSteps() {

        ArrayList<ArrayList<Integer>> s = new ArrayList();

        for (int i = 0; i < project.steps.size() - 1; i++) {

            Step step     = project.steps.get(i    );
            Step nextStep = project.steps.get(i + 1);

            int stepSliceCount = step.duration / (MainIOIOActivity.PERIOD / 1000);

            Log.i("~~~~~~~~~~~~~~~~", "Step duration=" + step.duration + " ms, slices=" + stepSliceCount);

            for (int sliceIndex = 0; sliceIndex < stepSliceCount; sliceIndex++) {

                ArrayList<Integer> interpolatedPositions = new ArrayList();

                for (int positionIndex = 0; positionIndex < step.positions.size(); positionIndex++) {

                    int initialPosition = step.positions.get(positionIndex);
                    int endPosition     = nextStep.positions.get(positionIndex);

                    // Interpolate the position
                    float deltaPosition = endPosition - initialPosition;
                    float deltaPerSlice = deltaPosition / stepSliceCount;

//                    if (sliceIndex == 0)
//                        Log.i("~~~~~~~~~~~~~~~~", "S" + positionIndex + " [" + initialPosition + " -> " + endPosition + "] delta: " + deltaPosition + ", deltaPerSlice: " + deltaPerSlice);

                    Float interpolatedPercentage = initialPosition + deltaPerSlice * sliceIndex;
                    Float interpolatedPosition = MainIOIOActivity.MIN + interpolatedPercentage * MainIOIOActivity.PERCENT;

                    interpolatedPositions.add(interpolatedPosition.intValue());
                }

                if (sliceIndex % 10 == 0)
                    Log.i("~~~~~~~~~~~~~~~~", "slice " + sliceIndex + ": " + interpolatedPositions);

                s.add(interpolatedPositions);
            }

            Log.i("~~~~~~~~~~~~~~~~", "");


        }

        if (s.size() != sliceCount)
            throw new RuntimeException("Expecting " + sliceCount + " slices but got " + s.size());

        return s;

    }

}


